systemd-hosts.d.service

```diff
@@ -3,13 +3,66 @@ Description=Keep your hosts records in logically separated files
 After=local-fs.target
 
 [Service]
+# Security
+ProtectSystem=yes
+ProtectHome=yes
+PrivateDevices=yes
+ProtectKernelTunables=yes
+ProtectKernelModules=yes
+ProtectControlGroups=yes
+ProtectKernelLogs=yes
+SystemCallFilter=@system-service
+SystemCallErrorNumber=EPERM
+NoNewPrivileges=yes
+PrivateTmp=yes
+LockPersonality=yes
+MemoryDenyWriteExecute=yes
+ProtectClock=yes
+RestrictNamespaces=yes
+RestrictRealtime=yes
+RestrictSUIDSGID=yes
+RuntimeDirectory=systemd/hosts.d
+RuntimeDirectoryPreserve=yes
+
+# Service
 Type=oneshot
+
+# Environment
 Environment="HOSTS_FILE=/etc/hosts"
 Environment="HOSTS_DIRECTORY=/etc/hosts.d"
-ExecStart=/bin/sh -c 'echo -e "## Generated by systemd-hosts.d ##\\n" > $HOSTS_FILE'
-ExecStart=/bin/sh -c 'echo -e "## All changes in this file will be ignored ##\\n" >> $HOSTS_FILE'
-ExecStart=/bin/sh -c 'cat $HOSTS_DIRECTORY/*.conf | grep \'^[[:blank:]]*[^[:blank:]#;]\' >> $HOSTS_FILE'
-ExecStart=/bin/chmod ugo=r $HOSTS_FILE
+# Loosely match IPv4 or IPv6 followed by hostname and hostalias(s)
+Environment="HOSTS_REGEX_BASE=^\\([0-9]\\{1,3\\}\\.\\|[0-9a-fA-F]\\{1,4\\}:\\|::\\).\\+\\?\\(\\s\\+[a-zA-Z0-9.-]\\{1,\\}\\)\\{1,\\}$"
+Environment="HOSTS_REGEX_COMMENT=^#.*$"
+# Override with HOSTS_ENABLE_COMMENT=yes to include comments from the source conf files
+Environment="HOSTS_ENABLE_COMMENT=no"
+# Override with HOSTS_ENABLE_BACKUP=no to disable hosts file backup creation
+Environment="HOSTS_ENABLE_BACKUP=yes"
+
+# Condition - validate that matching regex is found in the source before overwriting $HOSTS_FILE
+ExecCondition=/bin/sh -c 'grep "$HOSTS_REGEX_BASE" "$HOSTS_DIRECTORY"/*.conf >/dev/null'
+
+# Create a single backup
+ExecStartPre=/bin/sh -c 'if test "$HOSTS_ENABLE_BACKUP" = "yes"; then cp --dereference -av $HOSTS_FILE ${HOSTS_FILE}~; fi'
```

Comment:
First, if you intend to inline the shell expression anyways, the use of `&&`  or `||` operators will look that little bit clearer.

Second, could you please explain how does this help? At this point the hosts file is a duplicate of what's already in the hosts directory, because it's the pacman script's job to back up/restore the original hosts file when the packet is being installed/uninstalled.

Reply:
It's just a safeguard  mainly for the first destructive run. If something goes wrong, you still have the backup. It *feels* better to have at least a basic backup when automatically clobbering entire files. Ultimately it's unnecessary as you suggest.